features/timeseries/src/main/java/org/opennms/netmgt/timeseries/integration/persistence/MetaTagDataLoader.java

```diff
@@ -0,0 +1,184 @@
+        this.config = config;
+    }
+
+    public Map<String, String> load(final CollectionResource resource) {
+        return sessionUtils.withReadOnlyTransaction(() -> {
+
+            final Map<String, String> tags = new HashMap<>();
+            List<Scope> scopes = new ArrayList<>();
+
+            // resource related scope
+            scopes.add(getScopeForResource(resource));
+
+            // node related scopes
+            String nodeCriteria = getNodeCriteriaFromResource(resource);
+            Optional<OnmsNode> nodeOptional = getNode(nodeCriteria);
+            if(nodeOptional.isPresent()) {
+                OnmsNode node = nodeOptional.get();
+                scopes.add(this.entityScopeProvider.getScopeForNode(node.getId()));
+                scopes.add(this.entityScopeProvider.getScopeForAssets(node.getId()));
+                if (resource.getResourceTypeName().equals(CollectionResource.RESOURCE_TYPE_IF)) {
+                    // We expect #getInstance to return the ifIndex for interface-level resources
+                    try {
+                        int ifIndex = Integer.parseInt(resource.getInstance());
+                        scopes.add(this.entityScopeProvider.getScopeForInterfaceByIfIndex(node.getId(), ifIndex));
+                    } catch(NumberFormatException nfe) {
+                        // pass
+                    }
+                }
+                // We cannot retrieve service meta-data - resource time resources contain the IP address and service name, but not the node
+            }
+
+            // create tags for scopes
+            Scope scope = new FallbackScope(scopes);
+            Map<String, String> configuredMetaTags = this.config.getConfiguredMetaTags();
+            for(Map.Entry<String, String> entry: configuredMetaTags.entrySet()) {
+                final String value = Interpolator.interpolate(entry.getValue(), scope);
+                // Ignore tags with empty values
+                if (Strings.isNullOrEmpty(value)) {
+                    continue;
+                }
+                tags.put(entry.getKey(), value);
+            }
+
+            // create tags for categories
+            nodeOptional.ifPresent(onmsNode -> mapCategories(tags, onmsNode));
+            return tags;
+        });
+    }
+
+    private void mapCategories(final Map<String, String> tags, final OnmsNode node) {
+        Objects.requireNonNull(node);
+        if(config.isCategoriesEnabled()) {
+            node.getCategories().stream()
+                    .map(OnmsCategory::getName)
+                    .forEach(catName -> tags.put("cat_" + catName, catName));
+        }
+    }
+
+    public Scope getScopeForResource(final CollectionResource resource) {
+        if (resource == null) {
+            return EmptyScope.EMPTY;
+        }
+        return new ObjectScope<>(resource)
+                .map("resource", "criteria", (r) -> Optional.ofNullable(getNodeCriteriaFromResource(resource)))
```

Comment:
Let's store this in the "node" context instead (and move it to the EntityScopeProvider).

The criteria should also be derived from the OnmsNode object and not the resource. (It's possible that the resource only references the node id, but an FS:FID is present on the node.)

Reply:
done.